### PR TITLE
Fix emitting super-call when using prologue directives

### LINKED ISSUE #6901: if first line in constructor is 'somestring', "this" is called before calling super() in constructor
Hi,

we have experienced a problem where `this` is called before `super()` in a constructor function. This happen if the first line in a constructor is a string (for e.g `ngInject`;) 

Typescript
```typescript
class A  {
	blub = 6;
}


class B extends A {
  
  blub = 12;
  
  constructor() {
	'someStringForEgngInject';
    super()
  }
}

console.log(new B().blub);
```

and here the transpiled output
```js
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var A = (function () {
    function A() {
        this.blub = 6;
    }
    return A;
})();
var B = (function (_super) {
    __extends(B, _super);
    function B() {
        'someStringForEgngInject';
        this.blub = 12;
        _super.call(this);
    }
    return B;
})(A);
console.log(new B().blub);

```

The transpiled output with `'someStringForEgngInject'` removed works correct.

```js
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var A = (function () {
    function A() {
        this.blub = 6;
    }
    return A;
})();
var B = (function (_super) {
    __extends(B, _super);
    function B() {
        _super.call(this);
        this.blub = 12;
    }
    return B;
})(A);
console.log(new B().blub);
```

We are using Typescript 1.7.5 and compile to ES6 which is handed of to Babel which will print a error message like this
```
Module build failed: SyntaxError: C:/dev/a.js: 'this' is not allowed before super()
```

The code for transpiling constructors probably "assumes" that the first line is always the "super" call and insert all default values in the second line?!

Issue found by @loxy, CC'ing @davidreher

Cheers,
Fabian

### PR DESCRIPTION
**Pull request by yuit**
Fix #6901

### CODE REVIEW

File: src/compiler/emitter.ts [hunk 1]
```diff
@@ -4766,9 +4766,14 @@ const _super = (function (geti, seti) {
                 emitToken(SyntaxKind.CloseBraceToken, body.statements.end);
             }
 
-            function findInitialSuperCall(ctor: ConstructorDeclaration): ExpressionStatement {
+            /**
+             * Return the statement at a given index if it is a super-call statement
+             * @param ctor constructor declaration
+             * @param index an index to constructor's body to check
+             */
+            function getSuperCallAtGivenIndex(ctor: ConstructorDeclaration, index: number): ExpressionStatement {
```
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: returns CallExpression

Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: I'd just take a NodeArray<Statements>?

Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: Can you add an explicit return at the end?

Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: Why doesn't this just return a boolean if there's a super call at that index

File: src/compiler/emitter.ts [hunk 2]
```diff
@@ -4766,9 +4766,14 @@ const _super = (function (geti, seti) {
...
+             * @param ctor constructor declaration
+             * @param index an index to constructor's body to check
+             */
+            function getSuperCallAtGivenIndex(ctor: ConstructorDeclaration, index: number): ExpressionStatement {
                 if (ctor.body) {
-                    const statement = (<Block>ctor.body).statements[0];
+                    const statement = (<Block>ctor.body).statements[index];
                     if (statement && statement.kind === SyntaxKind.ExpressionStatement) {
                         const expr = (<ExpressionStatement>statement).expression;
                         if (expr && expr.kind === SyntaxKind.CallExpression) {
```
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]:
```ts
return !!expr && isSuperCallExpression(expr);
```

Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: Actually, `expression` isn't optional, you should be able to write `isSuperCallExpression(expr)`.

File: src/compiler/emitter.ts [hunk 3]
```diff
@@ -4766,9 +4766,14 @@ const _super = (function (geti, seti) {
...
             }
 
-            function findInitialSuperCall(ctor: ConstructorDeclaration): ExpressionStatement {
+            /**
+             * Return the statement at a given index if it is a super-call statement
+             * @param ctor constructor declaration
+             * @param index an index to constructor's body to check
+             */
+            function getSuperCallAtGivenIndex(ctor: ConstructorDeclaration, index: number): ExpressionStatement {
                 if (ctor.body) {
```
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: If you invert the conditions, this becomes a lot cleaner.

`if (!statements || index >= statements.length) {
    return false;
}

const statement = statements[index];
if (statement.kind !== SyntaxKind.ExpressionStatement) {
    return false;
}

return isSuperCallExpression((statement as ExpressionStatement).expression);

**Comment by DanielRosenwasser** [comment]
Is it guaranteed that the initial super call will be an expression statement? Is it not possible to have the following?

```ts
1 && super()
```

**Comment by mhegazy** [comment]
yes, otherwise we error. we only skip prologues (e.g. "use strict").

**Comment by mhegazy** [comment]
:+1:

**Comment by mhegazy** [comment]
please port to master as well.

**Comment by yuit** [comment]
@DanielRosenwasser  yep it will an error. There is a discussion/issue whether we should allow that :) #6908

File: src/compiler/emitter.ts (hunk 1 above)
Comment (yuit) [inline src/compiler/emitter.ts]: We can't just return boolean because we should the super-call found here for emitting later

**Pull request merged**